### PR TITLE
Use requester argument of request.defaults

### LINKED ISSUE #72: Use requester argument of request.defaults
Using the requester argument of `request.defaults` will resolve some issues and support all request methods.

This should fix:
* #48 
* #58 
* #64 
* Possibly #44 

Those issues can be closed when this lands and are stale anyway.

### PR DESCRIPTION
**Pull request by pro-src**
Closes #72 

The tests must be rewritten because you can't stub `request.get` or `request.post` the same way as before.

### CODE REVIEW

**Comment by pro-src** [comment]
@codemanki It's starting to look like you've abandoned this project. I'm done sending PRs until you respond properly. I'll just republish this instead.

File: index.js [hunk 1]
```diff
@@ -1,94 +1,59 @@
...
+module.exports = defaults.call(requestModule, {
+  // Cookies should be enabled
+  jar: requestModule.jar(),
+  headers: {
+    'User-Agent': 'Ubuntu Chromium/34.0.1847.116 Chrome/34.0.1847.116 Safari/537.36',
+    'Cache-Control': 'private',
+    'Accept': 'application/xml,application/xhtml+xml,text/html;q=0.9, text/plain;q=0.8,image/png,*/*;q=0.5'
+  },
+  // Cloudflare requires a delay of 5 seconds, so wait for at least 6.
+  cloudfareTimeout: 6000,
```
Comment (pro-src) [inline index.js]: typo

File: index.js [hunk 2]
```diff
@@ -115,17 +80,17 @@ function processRequestResponse(options, requestResult, callback) {
 
   isChallengePresent = stringBody.indexOf('a = document.getElementById(\'jschl-answer\');') !== -1;
   isRedirectChallengePresent = stringBody.indexOf('You are being redirected') !== -1 || stringBody.indexOf('sucuri_cloudproxy_js') !== -1;
-  isTargetPage = !isChallengePresent && !isRedirectChallengePresent;
+  // isTargetPage = !isChallengePresent && !isRedirectChallengePresent;
```
Comment (pro-src) [inline index.js]: Unused variable. I commented it out.

File: index.js [hunk 3]
```diff
@@ -177,7 +141,7 @@ function solveChallenge(response, body, options, callback) {
     return callback({errorType: 3, error: 'I cant extract method from setTimeOut wrapper'}, response, body);
   }
 
-  challenge_pass = body.match(/name="pass" value="(.+?)"/)[1];
+  var challenge_pass = body.match(/name="pass" value="(.+?)"/)[1];
```
Comment (pro-src) [inline index.js]: This was an unintended global variable assignment.

**Comment by codemanki** [comment]
@pro-src hey! I didn't abandon it, just didn't have much time to look into your PRs. I promise to review all of them in the next couple days, sorry.

**Comment by pro-src** [comment]
He's alive! Glad to hear it.

File: index.js [hunk 4]
```diff
@@ -1,94 +1,62 @@
...
-};
-
-/**
- * Performs post request to url with headers.
- * @param  {String}        url
- * @param  {String|Object} body        Will be passed as form data
- * @param  {Function}      callback    function(error, response, body) {}
- * @param  {Object}        headers     Hash with headers, e.g. {'Referer': 'http://google.com', 'User-Agent': '...'}
- */
-cloudscraper.post = function(url, body, callback, headers) {
```
Comment (pro-src) [inline index.js]: Request has a formData option so I wasn't sure why this was being used. Is there some reason to prefer this over request's options?

**Comment by pro-src** [comment]
You can see the tests I've written on this branch: https://github.com/pro-src/cloudflare-scrape/tree/requester-tests

And the CI build is here: https://travis-ci.org/pro-src/cloudflare-scrape/builds/498635135

I've yet to rewrite the errors test file so it's excluded from that build. There's quite a few changes so I'll wait to update this PR until I have your feedback.

File: index.js (hunk 4 above)
Comment (codemanki) [reply index.js]: No special reason behind that, I didn't look thoroughly into request's api

File: index.js [hunk 5]
```diff
@@ -1,94 +1,62 @@
...
-  headers['Content-Type'] = headers['Content-Type'] || 'application/x-www-form-urlencoded; charset=UTF-8';
-  headers['Content-Length'] = headers['Content-Length'] || data.length;
 
-  performRequest({
-    method: 'POST',
-    body: data,
-    url: url,
-    headers: headers
-  }, callback);
+var originalDefaults = requestModule.defaults;
```
Comment (codemanki) [inline index.js]: Wow, this is very nice. I wasn't aware of this feature in request. Looks like it would make things much easier

**Comment by codemanki** [comment]
@pro-src would it make sense to create a new branch for these changes ( i.e `ver-3.0.0`), so we could both contribute to the work you are doing? There are still a couple of todo items in the readme that I would want to implement. And then once it is done, I will merge that branch into master. What do you think?

**Comment by pro-src** [comment]
Sounds great and necessary due to the breaking changes. If you want, you can go ahead and great the v3.0.0 branch and I'll update this PR to be for that branch instead of master. At that point, feel free to merge it.

**Comment by pro-src** [comment]
@codemanki, The tests are written and passing. I'm done with this PR if everything looks good to you.

**Comment by codemanki** [comment]
@pro-src done! https://github.com/codemanki/cloudscraper/tree/v3.0.0 Could you please rebase your PR on that branch. 
So as you said, I then merge your PR and continue working on it in that branch. Is this a good idea or you would do it somehow differently?

**Comment by pro-src** [comment]
@codemanki LGTM

**Comment by codemanki** [comment]
@pro-src I'm currently going through your code, and it looks like for an URL protected by cloudflare it throws an error because of the 503 status code:
```
cloudscraper.get('https://iload.to/');
```
results in 
```
Unhandled rejection StatusCodeError: 503 - {"type":"Buffer","data":[60,33
```
This is thrown by a request-promise library, It looks like we need to catch 503s as for us this is a successful response and handle them by passing the result to the callback or `processRequestResponse` . What do you think?

**Comment by pro-src** [comment]
@codemanki, a promise rejection should never occur before `processRequestResponse` performs it's callback. I think this implies that cloudscraper isn't seeing a challenge, captcha, or error with that 503, perform it's callback, and then request-promise rejects because of the status code. I'm still looking into this.

**Comment by pro-src** [comment]
@codemanki, I'm sorry that I didn't catch this until now but the callback we're setting on request is never called. The way the `test/helper` is calling back is different from the way the request actually calls back.

Before the Request constructor ever returns:
https://github.com/request/request/blob/master/request.js#L187-L188
```js
    self.on('error', self.callback.bind())
    self.on('complete', self.callback.bind(self, null))
```

The callback isn't being replaced until after the constructor returns when those event listeners are set. The complete or error event is emitted on the request instance and cloudscraper's callback is bypassed entirely.

`request-promise` sets it's callback during `Request.prototype.init` and I think we should follow suit. Also, the `test/helper` should emit the complete event on the request instance instead of calling `request.callback` to mimic the real request behavior. I'll send a PR to fix this.

**Pull request merged**